### PR TITLE
Windows: Support credential specs

### PR DESCRIPTION
**Pull request by lowenna**
Signed-off-by: John Howard jhoward@microsoft.com

Adds support for a credential spec to be passed through from either the registry or file to support emulated domain identity. @patricklang 

Usage example:

```
docker built -t test --security-opt credentialspec=file://foo.txt .
```

Or

```
docker run -it --rm --security-opt credentialspec=registry://foo nanoserver cmd
```

Where in the first example, `foo.txt` exists under the `docker root directory\credentialspecs` and contains the credential blob (defaults to `c:\programfiles\docker\credentialspecs\...`), and in the second example, the key `foo` exists under `HKLM\SOFTWARE\Microsoft\Windows NT\CurrentVersion\Virtualization\Containers\CredentialSpecs`, and similarly contains the credential blob. In both cases, the blob is the non-secret identifying part, the contents of which are orthogonal to this PR and to docker - it is managed outside of docker. Docker is opaque to its actual contents.

image

For CI, it requires the CI servers to be domain joined, so adding test cases for CI isn't realistic. We're working on that separately.

@NathanMcCauley

### CODE REVIEW

**Comment by crosbymichael** [comment]
@jhowardmsft don't you need a line in the hack/vendor.sh file with the commit for the change?

**Comment by jhowardmsft** [comment]
the vendor PR is going in separately covering a few changes.

**Comment by jhowardmsft** [comment]
Rebased after #23392 merged.

**Comment by crosbymichael** [comment]
On the design side i'm not sure if labels should be overloaded for this.  Is this the first time we are using labels like this or is it being use for another setting somewhere else?

**Comment by thaJeztah** [comment]
Wouldn't this be a task for an entrypoint script? e.g. `regedit.exe /s something.reg`, or won't that work?

**Comment by cpuguy83** [comment]
I agree with @crosbymichael.
Tend to say labels are for systems that sit above docker, not for docker's subsystems.

**Comment by jhowardmsft** [comment]
Label was used after a Microsoft/Docker conf call recommendation. Patrick - please fill in the gaps here.

**Comment by PatrickLang** [comment]
We also considered environment and --security-opts. The problem is that security_opts already has multiple unrelated things in there and we didn't want to contribute to that. Our discussion with @NathanMcCauley was that labels would be a better location, and that its clear its specific to Microsoft for now. Long term we want this to be set by an upper layer along the lines of the DockerCon EU Universal Control Plane prototype. That one used labels.

**Comment by crosbymichael** [comment]
Was a dedicated flag considered for this? Also is this path a path inside the container or on the host?  If it is on the host this should not be inside a docker at all as it makes images not portable.

**Comment by PatrickLang** [comment]
It's a path on the host, and its completely optional. The container shouldn't have this data stored in it. If its not provided, the container will use the default built-in Windows accounts the same way they did before.

**Comment by thaJeztah** [comment]
There's some earlier discussion about using labels for configuration; this was about using labels to configure networking https://github.com/docker/docker/pull/13476

**Comment by crosbymichael** [comment]
Yes, we should not use labels for this.  

I don't see a problem adding a flag and api field for this.  Maybe a simple `--credentials` or similar may work.  

However, I don't see how we can have this in the dockerfile.  Even if it is optional it makes the images build be non portable since this references a filepath on the host system that may not exist on all systems.  

To move this forward we should discuss a proper flag name that everyone is happy with and remove this feature from the Dockerfile.  Also these API fields  should live in the HostConfig and not the container's Config structure. 

If the Dockerfile requirement is something being that the containers building the image need these credentials to run successfully, we can have a `docker build` flag, the same way we do with the cgroup options to add the correct credentials during build time but ensuring they are not committed to the final result. 

Let me know if you have any questions.  Happy to help get this merged.

**Comment by jhowardmsft** [comment]
@crosbymichael PR submitted to match ^^. Will submit a PR to engine-api and update further once that is merged and possible to re-vendor. Now has the --credentialspec field to both docker run and docker build. No longer cached in the image.  Note that this is something we would really like as part of 1.12 to show at Dockercon if possible.

**Comment by crosbymichael** [comment]
@jhowardmsft thanks for updating this. I'll check it out

**Comment by jhowardmsft** [comment]
Design good now and can move forward to code review? (Appreciate we're all busy at Dockercon though :smile:)

**Comment by crosbymichael** [comment]
@jhowardmsft the vendor checks are still failing

**Comment by jhowardmsft** [comment]
@crosbymichael I'm waiting for https://github.com/docker/engine-api/pull/282 to be merged to fix up the vendoring side of this PR.

**Comment by icecrime** [comment]
Not to diminish the priority of the patch, but removing from the milestone as Windows is not yet aligned to the release process (meaning: we build from master).

**Comment by jhowardmsft** [comment]
@thaJeztah How can we get this and the engine API PR moved forward? I've got folks really pushing hard for this functionality to demo

**Comment by thaJeztah** [comment]
let me start with adding an "needs attention" label

**Comment by PatrickLang** [comment]
Any updates here? I have been running a private build and would like to be able to run off master again.

**Comment by cpuguy83** [comment]
For me, I need an understanding of what a credential spec is, what problem it's solving, and if that problem applies to more than just Windows.

**Comment by PatrickLang** [comment]
This does start-time binding of network credentials to a Windows container. The typical example is that it allows the app running in the container such as a ASP.net website to use an Active Directory (based on Kerberos) group-managed service account to authenticate to a SQL database. No username/password is stored in the container.

The credential spec includes the data needed to identify the right service account from an AD domain controller.

**Comment by cpuguy83** [comment]
This sounds pretty similar to RedHat's issues with getting RHEL certificates into the build environment so build containers can access RH repos.

**Comment by jhowardmsft** [comment]
Not at all. One is outside the container. This is inside the container.

**Comment by cpuguy83** [comment]
@jhowardmsft I'm not sure I understand how RHEL's case would be outside the container.

**Comment by jhowardmsft** [comment]
That was my understanding. But none the less, this is completely different...

**Comment by jhowardmsft** [comment]
@NathanMcCauley Can you weigh in here? This is as per the previous design discussions. We weren't able to move this forward at the maintainers meeting today, and are sitting in circling pattern. @crosbymichael's feedback has been addressed. Overall this is one of the key scenarios for Windows containers, so moving this forward is something we really would appreciate. :smile:

**Comment by jhowardmsft** [comment]
@PatrickLang Can you follow up with @friism on how to move this forward now 1.12 is out of the way - I'm out next week. @NathanMcCauley - any feedback here?

**Comment by PatrickLang** [comment]
@diogomonica - This is the PR we talked about a few months ago. Can we get this in now that v1.12 is done? It's the last thing needed for Windows users to be able to use Windows Active Directory service accounts to secure connections (eg: website to SQL) without storing passwords in config files. I'm happy to go over it in more detail and cover questions - ping me on Slack.

**Comment by icecrime** [comment]
Apparently there were discussions with @diogomonica and @NathanMcCauley prior to this PR being submitted, so I'll you figure this out.

**Comment by jhowardmsft** [comment]
Rebased now engine-api is back in docker.

**Comment by jhowardmsft** [comment]
Rebased again

**Comment by icecrime** [comment]
I fully agree with @crosbymichael that a dedicated (Windows-specific) flag for `build` and `run` is the right approach (we should hire this guy!).

Regarding the relationship with secrets management:

- I have the feeling that credentials passing on Windows could/should ultimately reuse the same workflow than any platform-agnostic secret, but secret management is not a thing in Docker today so we can't treat that as a blocker. I however would ask you to please consider deprecating `--credentialspec` in favor of regular secrets when that feature lands (and if possible of course).
- It's my understanding that the current secrets design includes a Dockerfile instruction as a mean to express the dependency of an image on the existence of a particular secret: this serves the purpose of allowing to inspect images for runtime dependencies. Again, when that feature lands, I think we should really consider what can be done to encompass that use case.

So I guess design LGTM, moving to code review.

PS: I'm very happy you're having phone calls and brunch on Sundays, but this is an open source project and GitHub is our source of truth. It's extremely frustrating to be asked for an opinion after having missed over half of the discussions, especially when the people involved don't bother replying after a record-breaking 7 pings.

File: daemon/oci_windows.go
Comment (icecrime) [inline daemon/oci_windows.go]: This is unused now.

File: daemon/oci_windows.go
Comment (icecrime) [inline daemon/oci_windows.go]: Maybe this whole block would be simpler using `strings.HasPrefix` and `strings.TrimPrefix`? You wouldn't have to deal with the length of the resulting split string.

File: daemon/oci_windows.go
Comment (icecrime) [inline daemon/oci_windows.go]: Is it really just a warning? I wonder if we're not better of failing here that waiting for a potential (likely?) runtime failure as credentials happen to be missing.

File: daemon/oci_windows.go
Comment (icecrime) [inline daemon/oci_windows.go]: Same comment as above.

File: runconfig/hostconfig_solaris.go
Comment (icecrime) [inline runconfig/hostconfig_solaris.go]:
```
if hc != nil && len(hc.CredentialsSpec) != 0 {
    return fmt.Errorf(...)
}
return nil
```

Comment (icecrime) [inline runconfig/hostconfig_solaris.go]: Also not sure if we can have this code only written once in a file for anything not Windows.

File: runconfig/opts/parse.go
Comment (icecrime) [inline runconfig/opts/parse.go]: :scream:

Comment (jhowardmsft) [inline runconfig/opts/parse.go]: Your new keyboard :smiling_imp: 

image

File: daemon/oci_windows.go
Comment (jhowardmsft) [inline daemon/oci_windows.go]: Good catch. Will remove.

File: daemon/oci_windows.go
Comment (jhowardmsft) [inline daemon/oci_windows.go]: Yup, agreed. Will rework.

File: daemon/oci_windows.go
Comment (jhowardmsft) [inline daemon/oci_windows.go]: This is a good question, one which I somewhat agonised about. I really am not sure what the right answer is, I'm still sitting on the fence. However, I can easily be persuaded that this is an error, which I'll update it to. @PatrickLang unless you have a strong opinion the other way?

File: daemon/oci_windows.go
Comment (jhowardmsft) [inline daemon/oci_windows.go]: Updating as per above comment.

File: runconfig/hostconfig_solaris.go
Comment (jhowardmsft) [inline runconfig/hostconfig_solaris.go]: Agree with update. Will be in next push.

File: runconfig/opts/parse.go
Comment (jhowardmsft) [inline runconfig/opts/parse.go]: But yes, will remove in next push.

File: runconfig/hostconfig_solaris.go
Comment (jhowardmsft) [inline runconfig/hostconfig_solaris.go]: The best way to have it only once is to have a common `ValidateCredentialSpec` function in `hostconfig.go` utilising runtime.GOOS in the condition. I've had pushback about doing this having maintainers generally preferring _platform.go files, although sometimes it makes sense to inline it here. Happy to move to `hostconfig.go` if you think it's worth doing here. Haven't done that for the next update though.

**Comment by ehazlett** [comment]
fwiw this looks fine to me (not a mainainter).  +1 from @icecrime comment that we should revisit the build secret implementation when ready.  i think that could solve this case but i agree it shouldn't be a blocker for this.

**Comment by jhowardmsft** [comment]
@icecrime - with the one exception I mentioned which I'm not sure is worth it(?), your comments have been addressed. Thanks for the review. I've also added a test to handle failure cases (`TestRunCredentialSpecFailures`), but as I mentioned in the original PR comment, adding end-to-end success CI test cases isn't viable with the amount of additional infrastructure which would be required.

**Comment by jhowardmsft** [comment]
@icecrime @ehazlett And yes, I concur fully that once a generic implementation is available, this should be ported across to that solution. Happy to be on point for making that happen :smile:

**Comment by icecrime** [comment]
Thanks @jhowardmsft! LGTM

**Comment by jhowardmsft** [comment]
Rebased on master

**Comment by thaJeztah** [comment]
Needs another rebase already :sweat_smile:

**Comment by jhowardmsft** [comment]
And rebased again :smile:

**Comment by jhowardmsft** [comment]
And again....

**Comment by diogomonica** [comment]
I have reservations about adding this flag:
- It's a windows only feature (the fact that there are Linux-only flags is not a good argument as to why this should exist)
- The PR allows changing the build-time behavior, which is something that doesn’t exist in Docker right now, and there are a lot of internal discussions on whether that should be allowed or not.
- This PR is effectively bootstraping secrets on windows. Granted, might be the NLTM secrets, and not certificates, tokens, etc, but they are secrets nonetheless. Seems like using non-OS specific secrets distribution mechanism would be ideal.

This said, given that we don't have an alternative solution right now and that this seems to be a blocker for a legitimate use-case, I'll LGTM.

**Comment by jhowardmsft** [comment]
@diogomonica There is one alternate I can think of, which is also advantageous in that we don't have to deprecate a field in the hostconfig once a generic solution is available. This is to overload --security-opt instead from the CLI. Currently there are no security options on Windows.

So something like `docker run --security-opt credentialspec=file://foo ...`

Thoughts?

**Comment by jhowardmsft** [comment]
Edit: Actually that wouldn't work as --security-opt isn't a field to build.

File: daemon/start_windows.go
Comment (mavenugo) [inline daemon/start_windows.go]: minor nit : why getCS ? `getCredentialSpec` is short enough :)

File: runconfig/hostconfig_unix.go
Comment (mavenugo) [inline runconfig/hostconfig_unix.go]: minor nit : copy + paste error on Solaris :)

File: runconfig/hostconfig_windows.go
Comment (mavenugo) [inline runconfig/hostconfig_windows.go]: minor nit : another platform mix up.

File: runconfig/hostconfig_windows.go
Comment (mavenugo) [inline runconfig/hostconfig_windows.go]: where is the validation done and how ?

**Review by mavenugo — COMMENTED** [review]
Few minor nits and an open question.
It seems the changes are client/api changes, while the actual validation is out of scope (as explained in your description).

Since this falls under the secrets mgmt, I am not fully qualified to give comment on the design / requirements. But the code seems reasonable.

File: daemon/start_windows.go
Comment (jhowardmsft) [inline daemon/start_windows.go]: Sure, updated. Will be in next push.

File: runconfig/hostconfig_unix.go
Comment (jhowardmsft) [inline runconfig/hostconfig_unix.go]: Fixed in next push.

File: runconfig/hostconfig_windows.go
Comment (jhowardmsft) [inline runconfig/hostconfig_windows.go]: Fixed in next push.

File: runconfig/hostconfig_windows.go
Comment (jhowardmsft) [inline runconfig/hostconfig_windows.go]: This is the initial validation where the hostconfig gets passed from the caller into the daemon. The function as you can see is a no-op on Windows. The main point of this function is for similar things  such as `ValidateQoS` or `ValidateIsolation` in `hostconfig_unix.go` to error out if the daemon doesn't support those options. The validation of the value passed is done in `getLibcontainerdCreateOptions` in the update in this PR.

**Comment by jhowardmsft** [comment]
@icecrime @diogomonica Putting aside the other argument, I have purposefully in a second commit (will squash if acceptable) moved to using the existing security options in the `hostconfig` structure, rather than extending the structure for a field that will be deprecated once a central secrets management solution is in place in the daemon.  This will avoid backwards compatibility in removing the field, which I feel is important for something short-term.

Functionally from the first commit, it's idential.  The difference now is the credential spec is passed as a --security-opt in the CLI such as:

`docker run --it --rm --security-opt credentialspec=file://foo.txt nanoserver cmd`

If this options is attempted against a Linux daemon, the daemon will reject the request.

**Comment by cpuguy83** [comment]
Panic in test:

<details>
```
00:37:01 PANIC: docker_cli_run_test.go:4513: DockerSuite.TestRunCredentialSpecFailures
00:37:01 
00:37:01 ... Panic: runtime error: invalid memory address or nil pointer dereference (PC=0x45B283)
00:37:01 
00:37:01 C:/go/src/runtime/asm_amd64.s:479
00:37:01   in call32
00:37:01 C:/go/src/runtime/panic.go:458
00:37:01   in gopanic
00:37:01 C:/go/src/runtime/panic.go:62
00:37:01   in panicmem
00:37:01 C:/go/src/runtime/signal_windows.go:161
00:37:01   in sigpanic
00:37:01 docker_cli_run_test.go:4525
00:37:01   in DockerSuite.TestRunCredentialSpecFailures
00:37:01 C:/go/src/runtime/asm_amd64.s:479
00:37:01   in call32
00:37:01 C:/go/src/reflect/value.go:434
00:37:01   in Value.call
00:37:01 C:/go/src/reflect/value.go:302
00:37:01   in Value.Call
00:37:01 c:/gopath/src/github.com/docker/docker/vendor/src/github.com/go-check/check/check.go:816
00:37:01   in suiteRunner.forkTest.func1
00:37:01 c:/gopath/src/github.com/docker/docker/vendor/src/github.com/go-check/check/check.go:672
00:37:01   in suiteRunner.forkCall.func1
00:37:01 C:/go/src/runtime/asm_amd64.s:2086
00:37:01   in goexit
```
</details>

**Comment by jhowardmsft** [comment]
Thanks will look tomorrow - vacation day today.

**Comment by jhowardmsft** [comment]
Stupid test bug 👿 Fixed and verified locally.

**Comment by jhowardmsft** [comment]
And rebased. Still in two commits pending feedback as to whether this is a better option.

**Comment by cpuguy83** [comment]
Using `SecurityOpt` seems fine to me, especially since we are erroring out completely on non-Windows for build.

**Comment by diogomonica** [comment]
I also have nothing against it.

File: api/server/router/build/build_routes.go
Comment (cpuguy83) [inline api/server/router/build/build_routes.go]: Should be BadRequest or Conflict error.
And probably skip the flag and just use `security opts` since this is the daemon?

**Comment by jhowardmsft** [comment]
Based on last two comments, squashed the commits.

File: daemon/start_windows.go
Comment (cpuguy83) [inline daemon/start_windows.go]: No need to stat here, should be able to pull `os.IsNotExist` from `ioutil.ReadFile`

File: integration-cli/docker_cli_run_test.go
Comment (cpuguy83) [inline integration-cli/docker_cli_run_test.go]: I'd expect a different error here since it is prefixed with `file://`, just that there is no value.

File: integration-cli/docker_cli_run_test.go
Comment (cpuguy83) [inline integration-cli/docker_cli_run_test.go]: Any way to test a positive case?

File: daemon/start_windows.go
Comment (cpuguy83) [inline daemon/start_windows.go]: What about other error kinds?

**Review by cpuguy83 — CHANGES_REQUESTED** [review]

File: api/server/router/build/build_routes.go
Comment (jhowardmsft) [inline api/server/router/build/build_routes.go]: Hmm. Not sure about that. This is consistent with https://github.com/docker/docker/blob/master/api/server/router/build/build_routes.go#L65. And has to be picked up from the options (via `r.Form["securityopt"] a few lines higher).

File: daemon/start_windows.go
Comment (jhowardmsft) [inline daemon/start_windows.go]: Good point. Will be updated in next push.

File: daemon/start_windows.go
Comment (jhowardmsft) [inline daemon/start_windows.go]: Fixed in next push.

File: integration-cli/docker_cli_run_test.go
Comment (jhowardmsft) [inline integration-cli/docker_cli_run_test.go]: Not in the current CI infrastructure. To validate end-to-end, this requires an active directory domain and other infrastructure services.

File: integration-cli/docker_cli_run_test.go
Comment (jhowardmsft) [inline integration-cli/docker_cli_run_test.go]: Fair. Updated in next push.

**Comment by jhowardmsft** [comment]
@cpuguy83 Updated and pushed.

**Comment by jhowardmsft** [comment]
Rebased

File: daemon/start_windows.go
Comment (cpuguy83) [inline daemon/start_windows.go]: sure would be nice if there was a well tested helper function for grabbing k/v pairs out of SecuritryOpt... or rather `[]string`

File: api/server/router/build/build_routes.go
Comment (cpuguy83) [inline api/server/router/build/build_routes.go]: I still think we should fix this

**Review by cpuguy83 — APPROVED** [review]
As I don't have a Windows domain controller lying around to test this with, I trust this is well tested at MS :)

LGTM

File: api/server/router/build/build_routes.go
Comment (jhowardmsft) [inline api/server/router/build/build_routes.go]: Can you give me an example of what you were thinking? Can always do in a follow-up.

Comment (cpuguy83) [inline api/server/router/build/build_routes.go]: For the status code, just https://github.com/docker/docker/blob/master/api/errors/errors.go#L21

**Comment by jhowardmsft** [comment]
Rebased

**Comment by icecrime** [comment]
Still LGTM,

**Comment by cpuguy83** [comment]
Moving to docs review.

**Comment by jhowardmsft** [comment]
@thaJeztah Updated `build.md` and `run.md` (note there didn't seem to be anything at all in run.md saying about security options on Linux, but adding that is a bit beyond this PR. Just FYI). Anything else I should change too?

Note I've deliberately left it light-weight as to how to configure the credential specs and so on. That would be an entire white-paper in its own right, and something which will eventually be in the Microsoft documentation. @PatrickLang I assume that's something you and @neilpeterson are going to be taking care of. Once you have a link, I could hyperlink the docker docs to it.

**Review by vdemeester — APPROVED** [review]
docs LGTM 🐸

**Comment by soanni** [comment]
hello Docker Team, how can i try this PR? does the credentialspecs option is working in Windows 1.12.2-cs2-ws-beta ??
If not how can i get the nightly build or alpha whatever to try it.
Thank you in advance for your answer.

**Comment by jhowardmsft** [comment]
@PatrickLang Can you answer this? See email also.

@soanni Yes, this PR is in the latest CS build too.

**Comment by PatrickLang** [comment]
These docs are in review, I expect them to be on http://aka.ms/windowscontainers by end of week. I'll post a link when they're live.

**Comment by soanni** [comment]
@PatrickLang @jhowardmsft  thank you a lot!!

**Comment by soanni** [comment]
@PatrickLang any chance to take a look at these docs this week?

**Comment by PatrickLang** [comment]
@soanni got a review done today and pushed it live: https://msdn.microsoft.com/en-us/virtualization/windowscontainers/management/manage_serviceaccounts

The doc source and issue tracking are in this repo: https://github.com/microsoft/virtualization-documentation

**Comment by mpwagne** [comment]
This will help us a lot to containerize some of our products where we use windows auth for external resources (sql/shares). 

I have a couple of questions that I can't seem to find the definitive answer to. Hopefully someone can clarify for me...?

- I see we need min of Server 2012 on the domain controllers but what min version of the daemon do we need to support this? 
- Do you have a good example of the structure of the file content itself?
- The doc doesn't seem to go into details of the full credentialspec value to use. Is this correct?
   docker run ... --security-opt credentialspec=file://gMSA_Prod.txt myimage powershell
   (and the gMSA_Prod.txt file would be expected to be present on the host under the %programdata%\docker\credentialspecs  folder).
- Compose support for credentialspec...?
- Does this feature help us with the opposite flow, inbound access to an iis app running in the container using windows auth for example?  Or accessing a SQL server in a container using windows auth (mainly for local dev setup help)? 



By the way, I really enjoyed reviewing the history of this important PR. Shows some really good back and forth collaboration between Docker and MS folks.  :-)

**Comment by PatrickLang** [comment]
@mpwagne re: AD schema - it's in the [gMSA page](https://technet.microsoft.com/en-us/library/jj128431(v=ws.11).aspx#BKMK_gMSA_Req) linked from my document "The Active Directory schema in the gMSA domain’s forest needs to be updated to Windows Server 2012 to create a gMSA."

I'll track the other open Windows doc issues in my repo since they're questions on how Windows Server 2016 works. They'll show up linked here.

I think Docker-Compose should already work since `security_opt` is supported. If it doesn't work, can you open an issue on docker/compose and tag me on it?

**Comment by mpwagne** [comment]
@PatrickLang ok thanks. I will keep watch for more info.

What is the minimum version of the docker daemon that I need to get this support?

On the Compose, I quickly tried security_opt before and it complained but I was also struggling with syntax. I will create an issue if I still see issues after getting this working with a basic run command (after you fill me in on the syntax/file structure details). 

It might be a while to get that far though, I am still working with folks to get a 2012 based AD Schema that I can use. We seem to be behind a bit...  might have to spin up an isolated test AD myself.  This will be an uphill climb to get everyone on board with this 2012 schema dependency when we are ready to go up the environments to production. :-o

**Comment by dimitarivanov** [comment]
@mpwagne Compose supports this option
```
version: '2'
services:
  sqlserver:
    ...
    security_opt:
      - "credentialspec=file://sqlserverapp.json"
```
Docker version:
```
PS C:\> docker version
Client:
 Version:      1.12.2-cs2-ws-beta
 API version:  1.25
 Go version:   go1.7.1
 Git commit:   050b611
 Built:        Tue Oct 11 02:35:40 2016
 OS/Arch:      windows/amd64

Server:
 Version:      1.12.2-cs2-ws-beta
 API version:  1.25
 Go version:   go1.7.1
 Git commit:   050b611
 Built:        Tue Oct 11 02:35:40 2016
 OS/Arch:      windows/amd64
```
Docker Compose version:
```
PS C:\> docker-compose.exe version
docker-compose version 1.8.1, build 004ddae
docker-py version: 1.10.3
CPython version: 2.7.12
OpenSSL version: OpenSSL 1.0.2h  3 May 2016
```

**Pull request merged**
